server/src/test/java/org/spine3/server/procman/IteratingCommandRouterShould.java

```diff
@@ -0,0 +1,148 @@
+
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Set;
+
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.spine3.protobuf.Values.newStringValue;
+
+/**
+ * @author Alexander Yevsyukov
+ */
+public class IteratingCommandRouterShould {
+
+    private final TestCommandFactory commandFactory = TestCommandFactory.newInstance(getClass());
+
+    /** The object we test. */
+    private IteratingCommandRouter router;
+
+    /** The command message we route. */
+    private Message sourceMessage;
+
+    /** The context of the command that we route. */
+    private CommandContext sourceContext;
+
+    /** Command messages to be sent. */
+    private final List<Message> messages = ImmutableList.<Message>of(
+            newStringValue("uno"),
+            newStringValue("dos"),
+            newStringValue("tres"),
+            newStringValue("cuatro")
+    );
+
+    @Before
+    public void setUp() {
+        final BoundedContext boundedContext = BoundedContext.newBuilder()
+                                                            .build();
+        final CommandBus commandBus = boundedContext.getCommandBus();
+        commandBus.register(new CommandDispatcher() {
+            @Override
+            public Set<CommandClass> getCommandClasses() {
+                return CommandClass.setOf(StringValue.class);
+            }
+
+            @Override
+            public void dispatch(Command request) throws Exception {
+                // Do nothing.
+            }
+        });
+
+        sourceMessage = newStringValue(getClass().getSimpleName());
+        sourceContext = commandFactory.createCommandContext();
+
+        router = new IteratingCommandRouter(commandBus, sourceMessage, sourceContext);
+        router.addAll(messages);
+    }
+
+    @Test
+    public void routeFirst() throws Exception {
+        final CommandRouted commandRouted = router.routeFirst();
+
+        // Check that the source command is stored.
+        final Command source = commandRouted.getSource();
+        assertEquals(sourceMessage, Commands.getMessage(source));
+        assertEquals(sourceContext, source.getContext());
+
+        // Test that only only one command was produced by `routeFirst()`.
+        assertEquals(1, commandRouted.getProducedCount());
+
+        // Test that there's only one produced command and it has correct message.
+        final Command produced = commandRouted.getProduced(0);
+        final StringValue commandMessage = Commands.getMessage(produced);
+        assertEquals(messages.get(0), commandMessage);
+
+        // Test that the produced command context has correct fields.
+        assertTrue(Commands.sameActorAndTenant(sourceContext, produced.getContext()));
+
+        // Test that the event contains messages to follow.
+        assertEquals(messages.size() - 1, commandRouted.getMessageToFollowCount());
```

Comment:
This test does not depend on another test. It depends on `routeFirst()` (which it calls) working correctly. It is safe to assume so because we have another test that verifies it. 
If `routeFirst()` breaks this test would break, but we do want to know this too.